diff -urN tcpdump-cvs20001101.org/configure.in tcpdump-cvs20001101/configure.in
--- tcpdump-cvs20001101.org/configure.in	Wed Nov  1 16:11:03 2000
+++ tcpdump-cvs20001101/configure.in	Wed Nov  1 16:22:07 2000
@@ -530,7 +530,7 @@
 ac_cv_ssleay_path=no
 incdir=no
 for dir in /usr /usr/local /usr/local/ssl /usr/pkg; do
-	if test -d $dir/lib -a -f $dir/lib/libcrypto.a; then
+	if test -d $dir/lib -a -f $dir/lib/libcrypto.so; then
 		ac_cv_ssleay_path=$dir
 	fi
 	if test -d $dir/include/ssleay -a -f $dir/include/ssleay/des.h; then
