--- tcpdump_3_5rel2/Makefile.in.wiget	Fri Jul 14 14:06:19 2000
+++ tcpdump_3_5rel2/Makefile.in	Fri Jul 14 14:08:35 2000
@@ -134,12 +134,12 @@
 
 install: force
 	[ -d $(DESTDIR)$(BINDEST) ] || mkdir -p $(DESTDIR)$(BINDEST)
-	$(INSTALL) -m 550 -o bin -g @V_GROUP@ $(PROG) \
+	$(INSTALL) -m 550 $(PROG) \
 	    $(DESTDIR)$(BINDEST)/$(PROG)
 
 install-man: force
 	[ -d $(DESTDIR)$(MANDEST)/man1 ] || mkdir -p $(DESTDIR)$(MANDEST)/man1
-	$(INSTALL) -m 444 -o bin -g bin $(srcdir)/$(PROG).1 \
+	$(INSTALL) -m 444 $(srcdir)/$(PROG).1 \
 	    $(DESTDIR)$(MANDEST)/man1/$(PROG).1
 
 lint: $(GENSRC) force
